src/main/java/dev/jbang/cli/Setup.java

```diff
@@ -42,75 +60,74 @@ public static int setup(boolean withJava, boolean fresh) throws IOException {
-		}
+		binDir.toFile().mkdirs();
+
 		String cmd = "";
 		// Permanently add Jbang's bin folder to the user's PATH
-		Util.infoMsg("Adding jbang to PATH...");
 		if (Util.isWindows()) {
-			// Create the command to change the user's PATH
-			String newPath = binDir + ";";
-			if (withJava) {
-				newPath += jdkHome.resolve("bin") + ";";
-			}
-			String env = "[Environment]::SetEnvironmentVariable('Path', '" + newPath + "' + " +
-					"[Environment]::GetEnvironmentVariable('Path', [EnvironmentVariableTarget]::User), " +
-					"[EnvironmentVariableTarget]::User)";
+			String env = "";
 			if (withJava) {
+				String newPath = jdkHome.resolve("bin") + ";";
+				env += " ; [Environment]::SetEnvironmentVariable('Path', '" + newPath + "' + " +
+						"[Environment]::GetEnvironmentVariable('Path', [EnvironmentVariableTarget]::User), " +
+						"[EnvironmentVariableTarget]::User)";
 				env += " ; [Environment]::SetEnvironmentVariable('JAVA_HOME', '" + jdkHome + "', " +
 						"[EnvironmentVariableTarget]::User)";
 			}
-			if (Util.isUsingPowerShell()) {
-				cmd = "{ " + env + " }";
+			if (force || needsSetup()) {
+				// Create the command to change the user's PATH
+				String newPath = binDir + ";";
+				env += " ; [Environment]::SetEnvironmentVariable('Path', '" + newPath + "' + " +
+						"[Environment]::GetEnvironmentVariable('Path', [EnvironmentVariableTarget]::User), " +
+						"[EnvironmentVariableTarget]::User)";
 			} else {
-				cmd = "powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command \"" + env + "\" & ";
-			}
-		} else {
-			boolean updated = false;
-			// Update shell startup scripts
-			Path bashRcFile = getHome().resolve(".bashrc");
-			if (bashRcFile.toFile().exists()) {
-				changeScript(binDir, jdkHome, bashRcFile);
-				updated = true;
+				Util.infoMsg("jbang/bin already available in PATH.");
 			}
-			Path zshRcFile = getHome().resolve(".zshrc");
-			if (zshRcFile.toFile().exists()) {
-				changeScript(binDir, jdkHome, zshRcFile);
-				updated = true;
+			if (!env.isEmpty()) {
+				Util.infoMsg("Adding jbang/bin to PATH...");
+				if (Util.isUsingPowerShell()) {
+					cmd = "{ " + env + " }";
+				} else {
+					cmd = "powershell -NoProfile -ExecutionPolicy Bypass -NonInteractive -Command \"" + env + "\" & ";
+				}
 			}
-			if (!updated) {
-				Util.errorMsg("Did not find .bashrc nor .zshrc. jbang not able to be setup.");
-				return EXIT_GENERIC_ERROR;
+		} else {
+			if (force || needsSetup() || withJava) {
+				// Update shell startup scripts
+				boolean changed = false;
+				Path bashRcFile = getHome().resolve(".bashrc");
+				changed = changeScript(binDir, jdkHome, bashRcFile) || changed;
+				Path zshRcFile = getHome().resolve(".zshrc");
+				changed = changeScript(binDir, jdkHome, zshRcFile) || changed;
+				if (changed) {
+					Util.infoMsg("Adding jbang/bin to PATH...");
+				}
+			} else {
+				Util.infoMsg("jbang/bin already available in PATH.");
 			}
 		}
+
 		if (Util.isWindows()) {
```

Comment:
when already installed it prints:

```
 jbang setup
[jbang] jbang/bin already available in PATH.
Please start a new Shell for changes to take effect
```

if already available in PATH a restart shouldn't be required should it ?

Reply:
Fixed